File: expensereport-scala/src/main/scala/com/nelkinda/training/ExpenseReport.scala
package com.nelkinda.training

import java.util.Date

class ExpenseType {
}

object ExpenseType {
    val DINNER = new ExpenseType()
    val BREAKFAST = new ExpenseType()
    val CAR_RENTAL = new ExpenseType()
}

class Expense(val `type`: ExpenseType, val amount: Int)


class ExpenseReport {
    def printReport(htmlMode: Boolean, expenses: List[Expense]) {
        var totalExpenses = 0
        var mealExpenses = 0
        if (htmlMode) {
            println("""<!DOCTYPE html>
                    |<html>
                    |<head>
                    |<title>Expense Report: ${new Date()}</title>
                    |</head>
                    |<body>
                    |<h1>Expense Report: ${new Date()}</h1>""".stripMargin)
        } else {
            println(s"Expense Report: ${new Date()}")
        }

        if (htmlMode) {
            println("""<table>
                    |<thead>
                    |<tr><th scope="col">Type</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>
                    |</thead>
                    |<tbody>""".stripMargin)
        }
        for (expense <- expenses) {
            if (expense.`type` == ExpenseType.DINNER || expense.`type` == ExpenseType.BREAKFAST) {
                mealExpenses += expense.amount
            }
            var expenseName = expense.`type` match {
                case ExpenseType.DINNER => "Dinner"
                case ExpenseType.BREAKFAST => "Breakfast"
                case ExpenseType.CAR_RENTAL => "Car Rental"
            }
            var mealOverExpensesMarker = if (expense.`type` == ExpenseType.DINNER && expense.amount > 5000 || expense.`type` == ExpenseType.BREAKFAST && expense.amount > 1000) "X" else " "
            if (htmlMode) {
                println(s"<tr><td>${expenseName}</td><td>${expense.amount}</td><td>${mealOverExpensesMarker}</td></tr>")
            } else {
                println(s"${expenseName}\t${expense.amount}\t${mealOverExpensesMarker}")
            }
            totalExpenses += expense.amount
        }
        if (htmlMode) {
            println(s"</tbody>");
            println(s"</table>");
        }

        if (htmlMode) {
            println(s"<p>Meal Expenses: ${mealExpenses}</p>")
            println(s"<p>Total Expenses: ${totalExpenses}</p>")
        } else {
            println(s"Meal Expenses: ${mealExpenses}")
            println(s"Total Expenses: ${totalExpenses}")
        }

        if (htmlMode) {
            println(s"</body>");
            println(s"</html>");
        }
    }
}

object Main {
    def main(args: Array[String]) {
        new ExpenseReport().printReport(true, List[Expense](
            new Expense(ExpenseType.BREAKFAST, 1000),
            new Expense(ExpenseType.BREAKFAST, 1001),
            new Expense(ExpenseType.DINNER, 5000),
            new Expense(ExpenseType.DINNER, 5001),
            new Expense(ExpenseType.CAR_RENTAL, 4),
        ))
    }
}

File: expensereport-lua/ExpenseReport.lua
ExpenseType = {
    DINNER = 1,
    BREAKFAST = 2,
    CAR_RENTAL = 3,
}

Expense = { type, amount }

function Expense:new(type, amount)
    o = {}
    setmetatable(o, self)
    self.__index = self
    o.type = type
    o.amount = amount
    return o
end

function printReport(expenses, htmlMode)
    mealExpenses = 0
    total = 0
    if htmlMode then
        print("<!DOCTYPE html>")
        print("<html>")
        print("<head>")
        print("<title>Expense Report: " .. os.date() .. "</title>")
        print("</head>")
        print("<body>")
        print("<h1>Expense Report: " .. os.date() .. "</h1>")
    else
        print("Expenses: " .. os.date())
    end

    if htmlMode then
        print("<table>")
        print("<thead>")
        print("<tr><th scope=\"col\">Type</th><th scope=\"col\">Amount</th><th scope=\"col\">Over Limit</th></tr>")
        print("</thead>")
        print("<tbody>")
    end
    for i, expense in ipairs(expenses) do
        if expense.type == ExpenseType.DINNER or expense.type == ExpenseType.BREAKFAST then
            mealExpenses = mealExpenses + expense.amount
        end
        expenseName = ""
        if expense.type == ExpenseType.DINNER then expenseName = "Dinner" end
        if expense.type == ExpenseType.BREAKFAST then expenseName = "Breakfast" end
        if expense.type == ExpenseType.CAR_RENTAL then expenseName = "Car Rental" end
        mealOverExpensesMarker = (expense.type == ExpenseType.DINNER and expense.amount > 5000 or expense.type == ExpenseType.BREAKFAST and expense.amount > 1000) and "X" or " "
        if htmlMode then
            print("<tr><td>" .. expenseName .. "</td><td>" .. expense.amount .. "</td><td>" .. mealOverExpensesMarker .. "</td></tr>")
        else
            print(expenseName .. "\t" .. expense.amount .. "\t" .. mealOverExpensesMarker)
        end
        total = total + expense.amount
    end
    if htmlMode then
        print("</tbody>")
        print("</table>")
    end

    if htmlMode then
        print("<p>Meal Expenses: " .. mealExpenses .. "</p>")
        print("<p>Total Expenses: " .. total .. "</p>")
    else
        print("Meal Expenses: " .. mealExpenses)
        print("Total Expenses: " .. total)
    end

    if htmlMode then
        print("</body>")
        print("</html>")
    end
end

expenses = {
    Expense:new(ExpenseType.DINNER, 5000),
    Expense:new(ExpenseType.DINNER, 5001),
    Expense:new(ExpenseType.BREAKFAST, 1000),
    Expense:new(ExpenseType.BREAKFAST, 1001),
    Expense:new(ExpenseType.CAR_RENTAL, 4),
}

printReport(expenses, true)
